lisa/sut_orchestrator/azure/transformers.py

```diff
@@ -0,0 +1,259 @@
+
+        node_runbook = schema.RemoteNode(
+            name=runbook.vm_name,
+            public_address=runbook.public_address,
+            port=runbook.public_port,
+            username=runbook.username,
+            password=runbook.password,
+            private_key_file=runbook.private_key_file,
+        )
+        node = RemoteNode(
+            runbook=node_runbook, index=0, logger_name=f"{self.type_name()}_vm"
+        )
+        node.features = Features(node, platform)
+        node_context = get_node_context(node)
+        node_context.vm_name = runbook.vm_name
+        node_context.resource_group_name = runbook.resource_group_name
+
+        node.set_connection_info_by_runbook()
+        node.initialize()
+
+        # prepare vm for exporting
+        wa = node.tools[Waagent]
+        wa.deprovision()
+        node.execute("export HISTSIZE=0", shell=True)
+
+        # stop the vm
+        startstop = node.features[StartStop]
+        startstop.stop()
+
+        return node
+
+    def _export_vhd(self, platform: AzurePlatform, virtual_machine: Any) -> str:
+        runbook: VhdTransformerSchema = self.runbook
+        compute_client = get_compute_client(platform)
+
+        # generate sas url from os disk, so it can be copied.
+        self._log.debug("generating sas url...")
+        location = virtual_machine.location
+        os_disk_name = virtual_machine.storage_profile.os_disk.name
+        operation = compute_client.disks.begin_grant_access(
+            resource_group_name=runbook.resource_group_name,
+            disk_name=os_disk_name,
+            grant_access_data=GrantAccessData(access="Read", duration_in_seconds=86400),
+        )
+        wait_operation(operation)
+        sas_url = operation.result().access_sas
+        assert sas_url, "cannot get sas_url from os disk"
+
+        self._log.debug("getting or creating storage account and container...")
+        # get vhd container
+        if not runbook.storage_account_name:
+            runbook.storage_account_name = get_storage_account_name(
+                subscription_id=platform.subscription_id, location=location, type="t"
+            )
+
+        check_or_create_storage_account(
+            credential=platform.credential,
+            subscription_id=platform.subscription_id,
+            account_name=runbook.storage_account_name,
+            resource_group_name=runbook.resource_group_name,
+            location=location,
+            log=self._log,
+        )
+        container_client = get_or_create_storage_container(
+            runbook.storage_account_name, runbook.container_name, platform.credential
+        )
+
+        path = _generate_vhd_path(container_client, runbook.file_name_part)
+        self._log.debug(f"copying vhd: {path}")
+        blob_client = container_client.get_blob_client(path)
+        operation = blob_client.start_copy_from_url(
+            sas_url, metadata=None, incremental_copy=False
+        )
+        while True:
```

Comment:
we may need add a timeout here in case azure platform hit some issue, then the loop will cost quite long time.

Reply:
good catch, will add it. How about 30 minutes? I saw it's about 3 minutes in same location for 30 GB.